pkg/nfs/statefulset.go

```diff
@@ -0,0 +1,167 @@
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// PVCNamePrefix is the prefix of the PVC names used by the NFS StatefulSet.
+	// The PVC names are of the format <prefix>-<statefulset-pod-name>.
+	// If the NFS server statefulset name is "example-nfsserver" and the pod is
+	// named "example-nfsserver-0", the PVC will be named
+	// "nfs-data-example-nfsserver-0"
+	PVCNamePrefix = "nfs-data"
+)
+
+func (d *Deployment) createStatefulSet(size *resource.Quantity, nfsPort int, rpcPort int, metricsPort int) error {
+
+	replicas := int32(1)
+
+	// TODO: Check if the PVC already exists before attempting to create one.
+
+	ss := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            d.nfsServer.Name,
+			Namespace:       d.nfsServer.Namespace,
+			Labels:          d.nfsServer.Labels,
+			OwnerReferences: d.nfsServer.ObjectMeta.OwnerReferences,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			ServiceName: d.nfsServer.Name,
+			Replicas:    &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labelsForStatefulSet(d.nfsServer.Name, d.nfsServer.Labels),
+			},
+			Template:             d.createPodTemplateSpec(nfsPort, rpcPort, metricsPort, d.nfsServer.Labels),
+			VolumeClaimTemplates: d.createVolumeClaimTemplateSpecs(size, d.nfsServer.Labels),
+		},
+	}
+
+	return d.createOrUpdateObject(ss)
+}
+
+func (d *Deployment) createVolumeClaimTemplateSpecs(size *resource.Quantity, labels map[string]string) []corev1.PersistentVolumeClaim {
+	scName := d.nfsServer.Spec.GetStorageClassName()
+
+	claim := corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      PVCNamePrefix,
+			Namespace: d.nfsServer.Namespace,
+			Labels:    labels,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes:      []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			StorageClassName: &scName,
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{},
+			},
+		},
+	}
+
+	if size != nil {
+		claim.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceName(corev1.ResourceStorage): *size,
+		}
+	}
+
+	return []corev1.PersistentVolumeClaim{claim}
+}
+
+func (d *Deployment) createPodTemplateSpec(nfsPort int, rpcPort int, metricsPort int, labels map[string]string) corev1.PodTemplateSpec {
+
+	return corev1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: labelsForStatefulSet(d.nfsServer.Name, labels),
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					ImagePullPolicy: "IfNotPresent",
+					Name:            "nfsd",
+					Image:           d.nfsServer.Spec.GetContainerImage(),
+					// Args: []string{"nfs", "server", "--ganeshaConfigPath=" + NFSConfigMapPath + "/" + nfsServer.name},
+					Env: []corev1.EnvVar{
+						{
+							Name:  "GANESHA_CONFIGFILE",
+							Value: "/config/" + d.nfsServer.Name,
+						},
+					},
```

Comment:
Can you also add env vars for `NAME` and `NAMESPACE` set to the RWX PVC name and namespace?  This will be used for decorating the metrics.

Reply:
Exact PVC name won't be accurate. Because the volume claim is a template in StatefulSet, only the PVC name prefix is predictable. PVC name is appended by the StatefulSet name and a number for the replicas.
I guess the prefix name should be enough?

Reply:
It should be the name of the RWX vol, not the underlying StorageOS vol, so I think you can use `d.nfsServer.Name` and `d.nfsServer.Namespace` here.

Reply:
Added the env vars in this PR itself.